### ISSUE TITLE
Class not found exception LogginAdvice

### ISSUE THREAD

**Issue by Blacias** [raphw/byte-buddy#692]
Hi,
I am loading an Advice Class using URLClassLoader, but the app is throwing a ClassNotFound Exception exception wheneve I am using a class variable inside the function.

Agent : 
```java
Class<?> testClass=null;;
		try {
			testClass = Class.forName("com.exp.LoggingAdvice", true, classLoader);
		} catch (ClassNotFoundException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		System.out.println("Class = "+testClass.getCanonicalName());
		return new AgentBuilder.Default().disableClassFormatChanges()
				.with(AgentBuilder.RedefinitionStrategy.RETRANSFORMATION)
				.with(AgentBuilder.RedefinitionStrategy.REDEFINITION)
	            .with(AgentBuilder.TypeStrategy.Default.REDEFINE)
	            .type(ElementMatchers.hasSuperType(named(className)))
	            .transform(new AgentBuilder.Transformer.ForAdvice()
                        .include(classLoader)
                        .advice(named(methodName), testClass.getName())
                );
```

Output:

`Class = com.exp.LoggingAdvice

2019-07-25 13:10:41.932  INFO 17674 --- [nio-8080-exec-1] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring DispatcherServlet 'dispatcherServlet'
2019-07-25 13:10:41.933  INFO 17674 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Initializing Servlet 'dispatcherServlet'
2019-07-25 13:10:42.036  INFO 17674 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Completed initialization in 102 ms
LOL
2019-07-25 13:10:42.147 ERROR 17674 --- [nio-8080-exec-1] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Servlet execution threw an exception] with root cause

java.lang.ClassNotFoundException: com.exp.LoggingAdvice`

In the output, check 

> LOL

It is present in the same ClassFile(LoggingAdvice), but wherever there is a  First class variable in the function, it throws ClassNotFoundException.

Can anyone Help ? I am a beginner at this bytecode manipulation and unfortunately, to classloading concept aswell.

**Comment by raphw** [comment]
I assume that the advice class references classes that are not visible to the agent class loader.

You should not load the advice class, rather refer to it only by its name in the `advice` method. Byte Buddy will create a type representation based on all included class loaders to avoid such lookup failures.

**Comment by Blacias** [comment]
I appreciate the quick response.. 
The thing is, the advice works fine as long as I don't use the class variables. But when I try to use them, I get this ClassNotFoundException.

I am using Delegation to load my agent using a ClassLoader.

I guess I will try to look more into the ClassLoading like you suggested. 
Will update if I fix it somehow.

**Comment by raphw** [comment]
Is the exception thrown during the instrumentation or after the instrumentation is applied and your code is executed? If your target class loader does not know of the class, you need to make your auxiliary classes available on that class loader.

**Comment by Blacias** [comment]
> If your target class loader does not know of the class, you need to make your auxiliary classes available on that class loader

You were absolutely correct. In the advice class, I was trying to call methods, which ClassLoader had no idea where to look for. I wrote a CustomClassLoader class and pushed it into the main class loader. Now when I wanted to load something from the Advice, I call that CustomClassLoader class and load class from there. I hope using the customclassloader is better than pushing all my classes to the main class loader.

You were a big help already. Thank you so much.

I have a small doubt remaining, If you please.

I have a microservice which gets data from DB. I am trying to inject my agent on runtime.

If I inject my agent when no DB calls were made, My agent does not trace the data. On the other hand, if even a single call is made to DB from the microservice and after that I try to inject my agent. It traces the data.

I have come to the conclusion that maybe the class is needed to be initialized to be instrumented?

Will you please put a light on this? .... Thanks in advance.

**Comment by raphw** [comment]
I assume that you are probably loading your class during instrumentation what aborts the instrumentation as you cannot rely on a class while instrumenting it what causes a circularity.

Avoid referencing a class while you are instrumenting it.

**Comment by Blacias** [comment]
I am instrumenting java.sql.Statement class... I am sorry I did not understand what you tried to explain.

This is the code of my agent.

`return new AgentBuilder.Default().disableClassFormatChanges()
		.with(AgentBuilder.RedefinitionStrategy.RETRANSFORMATION)
		.with(AgentBuilder.TypeStrategy.Default.REDEFINE)
		.type(hasSuperType(named(className))) //className =  java.sql.Statement
		.transform(new AgentBuilder.Transformer.ForAdvice()
				.include(CustomClassLoader.getClassLoader())
				.advice(named(methodName), "com.exp.dbTracing.DBTracing")
                                //methodName = execute()
		                );
		         }`

**Comment by raphw** [comment]
This should not load the class, try registering a stream writing listener to see if an error is printed.

**Comment by Blacias** [comment]
I registered a stream writing listener but I did not see any error, all I could find was 
`IGNORE ....`
`COMPLETE ...`

The agent code is the same as above, just with the stream writing listener append to it.
I will paste my entire code here and maybe that could help better?

1. I have a loader agent which loads my bytebuddy agent in a different class loader : 

App.java //Loader agent
`public static void agentmain(String agentArgs, Instrumentation inst) {
		
		System.out.println( "Loading Class from jar!" );
		CustomClassLoader classLoader=null;
		try {
			classLoader = new CustomClassLoader( new URL[] {
					new URL("file:///home/user/Desktop/byte-buddy-agent-0.0.1.jar")},	
					Thread.currentThread().getContextClassLoader());
			CustomClassLoader.setClassLoader(classLoader);
                       //CustomClassLoader class to be pushed in app class loader
			classLoader.loadClass("demo.bytebuddy.logging.LoggingAgent")
			.getMethod("premainLoader", String.class, Instrumentation.class,URLClassLoader.class)
			.invoke(null,agentArgs,inst,classLoader);
			
		} catch (Exception e) {
			e.printStackTrace();
		}finally {
			try {
				classLoader.close();
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
		
	}`

2. LoggingAgent.java // internal agent

	public static void premainLoader(String agentArguments, Instrumentation instrumentation, URLClassLoader classloader) {
              createAgent("java.sql.Statement", "execute",cl).installOn(instrumentation);
	}

private static AgentBuilder createAgent(String className, String methodName, ClassLoader classLoader) {
		return new AgentBuilder.Default().disableClassFormatChanges()
				.with(AgentBuilder.RedefinitionStrategy.RETRANSFORMATION)
				//.with(AgentBuilder.RedefinitionStrategy.REDEFINITION)
	            .with(AgentBuilder.TypeStrategy.Default.REDEFINE)
	            .with(AgentBuilder.Listener.StreamWriting.toSystemError())
	            .type(hasSuperType(named(className)))
	            .transform(new AgentBuilder.Transformer.ForAdvice()
	            		.include(classLoader)
                        .advice(named(methodName), "com.exp.dbTracing.DBTracing")
                );
	            
	}



I am using bytebuddy agent loader to load this agent inside my application on runtine.

**Comment by raphw** [comment]
It looks correct to me. Is the class in question listed as ignored if you have not loaded it prior to the agent?

**Comment by Blacias** [comment]
This is weird, when the class is not loaded, ByteBuddy is not showing the class at all, neither in IGNORE nor in COMPLETE.

I am so sorry for taking this much of your time, I am trying to look what else I can do.

**Comment by mjok** [comment]
Had the same problem, i think here is the explanation:
You're instrumenting java.sql.Statement witch is in rt.jar and it's loaded with bootstrap classloader, witch doesn't know anything about your's Advice class'es.

There is a tool to load classes to bootstrap classloader  in Byte-Buddy, but in my case it's not working, cause I'am using another lib in my Advice, where package-private classes exist. 
(any solution here?)

**Comment by raphw** [comment]
That sounds right.

You could always move your entire agent to the bootstrap class path. It is not possible to reach any upward class loader from the boot loader.

**Comment by mjok** [comment]
Actually I have tried that.
If I do:
```
<...>
inst.appendToBootstrapClassLoaderSearch(new JarFile(new File("../target/probeAll.jar")));
<...>
```
where probeAll.jar is shaded jar, it does complain:


```
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:386)
	at sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:401)
Caused by: java.lang.LinkageError: loader constraint violation: when resolving interface method "net.bytebuddy.matcher.ElementMatcher$Junction.or(Lnet/bytebuddy/matcher/ElementMatcher;)Lnet/bytebuddy/matcher/ElementMatcher$Junction;" the class loader (instance of sun/misc/Launcher$AppClassLoader) of the current class, com/jkoolcloud/javaam/App, and the class loader (instance of <bootloader>) for the method's defining class, net/bytebuddy/matcher/ElementMatcher$Junction, have different Class objects for the type net/bytebuddy/matcher/ElementMatcher used in the signature
	at com.jkoolcloud.javaam.App.premain(App.java:63)
	... 6 more
```

**Comment by raphw** [comment]
That indicates that parts of your application was already loaded from the class path prior to the appending on the boot path such that some classes will be loaded from two loaders what makes them incompatible despite their nominal equality.

**Comment by Blacias** [comment]
I will try pushing my whole agent to bootstrap class loader then and try if it works. I do have some inhibition pushing the entire agent in the Bootstrap class loader. Will it be safe? I mean,  I was trying to keep my agent in a CustomClassLoader to keep the target app detached from it.

**Comment by raphw** [comment]
I'd recommend against it, too. It's a quick and dirty solution but it wirks most of the time. If you want to hide your agent from the user as much as possible, only push the necessary bits to the boot loader and isolate the rest somewhere else.

Ideally, you never distribute jars between different loaders.

**Comment by Blacias** [comment]
Well, I am writing it here because it may help someone.
This is in regard to instrument java.sql.Statement :

In my case, I wanted to instrument this when a database call is made for a mysql DB. 

First of all, bytebuddy can instrument java.sql.Statement class without pushing the agent into the Bootstrap class loader. I did it with the same code mentioned above. 

I debugged my code, added SystemOut listener and found out that, Bytebuddy was not able to find the Advice class which was to be triggered when a database call was made. The error was : 
`java.lang.IllegalStateException: Cannot resolve type description for com.example.DBTracing`

So what I did is, I put this class in the same class loader, where my agent was pushed. Please note that my Agentbuilder/ Transformer code is in custom class loader, but my advice is in System class loader, where my agent is loaded. And it worked!

**Comment by Blacias** [comment]
I have worked it out! Thanks for the help.

**Comment by Blacias** [comment]
Bytebuddy is able to instrument java.sql.Statement without pushing the agent into the Bootstrap Loader.
In my case what I wanted to check is, whenever a call is made from my java application to a mysql DB, I wanted my agent to trace it.

For that, I tried to instrument java.sql.Statement, but was facing some issues. So I added SystemOut listener and found out that Bytebuddy was not able to find my Advice class : 
`java.lang.IllegalStateException: Cannot resolve type description for com.exp.innerloader.DBTracing`

I am using an agentLoader which loads my main agent in a CustomClassLoader. So I added my advice files to the agentLoader, which in turn, loaded the advice class in SystemClassLoader and then, my agent was easily able to find the Advice class.
